plugins/network-device-injector/network-device-injector.go

```diff
@@ -0,0 +1,365 @@
+	verbose bool
+)
+
+// an annotated netdevice
+// https://man7.org/linux/man-pages/man7/netdevice.7.html
+type netdevice struct {
+	Name    string `json:"name"`     // name in the runtime namespace
+	NewName string `json:"new_name"` // name inside the pod namespace
+	Address string `json:"address"`
+	Prefix  int    `json:"prefix"`
+	MTU     int    `json:"mtu"`
+}
+
+func (n *netdevice) inject(nsPath string) error {
+	// Lock the OS Thread so we don't accidentally switch namespaces
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	link, err := netlink.LinkByName(n.Name)
+	if err != nil {
+		return err
+	}
+	ns, err := netns.GetFromPath(nsPath)
+	if err != nil {
+		return err
+	}
+	defer ns.Close()
+	// Devices can be renamed only when down
+	err = netlink.LinkSetDown(link)
+	if err != nil {
+		return err
+	}
+	// Save host device name into the container device's alias property
+	err = netlink.LinkSetAlias(link, link.Attrs().Name)
+	if err != nil {
+		return fmt.Errorf("fail to set alias for iface %s: %w", n.Name, err)
+	}
+	err = netlink.LinkSetNsFd(link, int(ns))
+	if err != nil {
+		return fmt.Errorf("fail to move link for iface %s to ns %d : %v", n.Name, int(ns), err)
+	}
+	// This is now inside the container namespace
+	err = netns.Set(ns)
+	if err != nil {
+		return fmt.Errorf("fail to set to ns %d: %v", int(ns), err)
+	}
+
+	link, err = netlink.LinkByName(n.Name)
+	if err != nil {
+		return err
+	}
+
+	err = netlink.LinkSetName(link, n.NewName)
+	if err != nil {
+		return err
+	}
+
+	if n.MTU > 0 {
+		err = netlink.LinkSetMTU(link, n.MTU)
+		if err != nil {
+			return err
+		}
+	}
+
+	if n.Address == "" {
+		return nil
+	}
+
+	nlAddr, err := netlink.ParseAddr(fmt.Sprintf("%s/%d", n.Address, n.Prefix))
+	if err != nil {
+		return err
+	}
+
+	err = netlink.AddrAdd(link, nlAddr)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// remove the network device from the Pod namespace and recover its name
+// Leaves the interface in down state to avoid issues with the root network.
+func (n *netdevice) release(nsPath string) error {
+	// Lock the OS Thread so we don't accidentally switch namespaces
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	ns, err := netns.GetFromPath(nsPath)
+	if err != nil {
+		return err
+	}
+	defer ns.Close()
+
+	// This is now inside the container namespace
+	err = netns.Set(ns)
+	if err != nil {
+		return fmt.Errorf("fail to set to ns %d: %v", int(ns), err)
+	}
+
+	link, err := netlink.LinkByName(n.NewName)
+	if err != nil {
+		return err
+	}
+
+	// Devices can be renamed only when down
+	err = netlink.LinkSetDown(link)
+	if err != nil {
+		return err
+	}
+
+	err = netlink.LinkSetName(link, n.Name)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// our injector plugin
+type plugin struct {
+	stub stub.Stub
+}
+
+func (p *plugin) RunPodSandbox(_ context.Context, pod *api.PodSandbox) error {
+	log.Infof("Started pod %s/%s...", pod.GetNamespace(), pod.GetName())
```

Comment:
nit (multiple occurrences): since we have a structured logger, it's nice to use fields

```suggestion
	log.WithField("namespace", pod.GetNamespace()).WithField("name", pod.GetName).Info("Started pod...")
```